packages/postgraphile-core/__tests__/kitchen-sink-schema.sql

```diff
@@ -461,6 +461,54 @@ create function c.return_table_without_grants() returns c.compound_key as $$
   select * from c.compound_key order by person_id_1, person_id_2 limit 1
 $$ language sql stable security definer;
 
+create function c.func_out(out o int) as $$
+  select 42 as o;
+$$ language sql stable;
+
+create function c.func_out_unnamed(out int) as $$
+  select 42;
+$$ language sql stable;
+
+create function c.func_out_setof(out o int) returns setof int as $$
+  select 42 as o
+  union
+  select 43 as o;
+$$ language sql stable;
+
+create function c.func_out_out(out first_out int, out second_out text) as $$
+  select 42 as first_out, 'out'::text as second_out;
+$$ language sql stable;
+
+create function c.func_out_out_unnamed(out int, out text) as $$
+  select 42, 'out'::text;
+$$ language sql stable;
+
+create function c.func_out_out_setof(out o1 int, out o2 text) returns setof record as $$
+  select 42 as o1, 'out'::text as o2
+  union
+  select 43 as o1, 'out2'::text as o2
+$$ language sql stable;
+
+create function c.func_in_out(i int, out o int) as $$
+  select i + 42 as o;
+$$ language sql stable;
+
+create function c.func_in_inout(i int, inout ino int) as $$
+  select i + ino as ino;
+$$ language sql stable;
+
+create function c.func_returns_table_one_col(i int) returns table (col1 int) as $$
+  select i + 42 as col1
+  union
+  select i + 43 as col1;
+$$ language sql stable;
+
+create function c.func_returns_table_multi_col(i int) returns table (col1 int, col2 text) as $$
+  select i + 42 as col1, 'out'::text as col2
+  union
+  select i + 43 as col1, 'out2'::text as col2;
+$$ language sql stable;
+
```

Comment:
We need to have some more trickier functions in here to really test this. Functions with output arguments that use compound types, output arguments that are tables (and check that relations still work on the tables); all these checks need to be done on all three types of functions: custom queries, custom mutations and computed fields.

I think we're also going to get problems for functions like the below that return raw `record` type without declaring the shape:

```sql
begin;
create function foo() returns record as $$ select 1, 'foo', 3; $$ language sql immutable;
select * from foo() as t(a1 int, a2 text, a3 int);
rollback;
```

These should be skipped by PostGraphile as we do not know how to specify the table alias required in the `select` statement. Adding them to the test suite will ensure that they are skipped, rather than causing PostGraphile to crash on startup.

Reply:
> output arguments that are tables (and check that relations still work on the tables)

Can you elaborate on your expectations for specifying tables as output arguments? Do you just want to support definitions like:

```sql
create function func_out_table3(out person) as $$
  ...
$$ language plpgsql stable;
```

These aren't currently exposed because of this: https://github.com/graphile/graphile-engine/blob/9d2855f2196424a5da1353c7c3390371d61593b1/packages/graphile-build-pg/src/plugins/PgQueryProceduresPlugin.js#L56-L63

..but that would be a simple to fix.

Reply:
In other words, if there's a single output argument of a table type (`func(out person)`), expose it just like `func() returns person`. Any other variation of output args that include a table type would be ignored. Example of function that would be ignored:
```
create function func_out_table99(out person, out something_else text)
  returns setof record as $$ ...
```
Correct?

Reply:
Oh you're right - things are way more complex than I realised - the codebase currently assumes that all arguments are input arguments (like that code you reference does) - we need to go through the current codebase and add "and is an input argument" checks in the relevant places. I wonder how computed columns should handle `inout` arguments... 🤔

You should be able to output a mix of scalar and compound types:

```sql
create type compound_type as (
  c int,
  d text
);
create table my_table (
  e int,
  f text
);

create function my_func(in a int, in b text, out x int, out y compound_type, out z my_table) as $$ ... $$ language sql;
```

Reply:
> we need to go through the current codebase and add "and is an input argument" checks in the relevant places

Done (I think) via 39a94ef

> if there's a single table output argument like `func(out person)`, expose it just like `func() returns person`

I added tests via f537a78  to ensure that this (and it's `setof` counterpart) work.

You want to go down the rabbit hole towards your `my_func` example within this PR?

Reply:
No apparent issues with compound types: 6e93503

Still working to get your `my_func` torture test to pass.

Reply:
Should also add a relation from my_table to another table and check that PostGraphile can navigate it as expected during the query.

Reply:
I think you’ll need to recurseDataGenerators on the record field of the mutation payload; can’t remember if that’s done automatically for you... might be?

Reply:
I’m happy to pitch in on this if you get stuck; the code to achieve it may be somewhat complex. It’s going to be a couple weeks before I can though.

Reply:
Ok, your `my_func` plus a related table is finally working: 8bf04ba. Hope I didn't butcher the implementation.. this is clearly the "make it work" phase. :sweat_smile:

Reply:
Connections are working now (c163c2e; test: 88d39aa). I realize this hijacking of PgScalarFunctionConnectionPlugin isn't ideal, but doing it this way helps me track the changes in logic. I can refactor after everything's working as intended.

Reply:
> I wonder how computed columns should handle inout arguments...
Any special concerns? I think it's working as expected: 06e111f

Reply:
Just wondering what'd happen if you made the table arg itself inout.

Reply:
It Just Works :tada: 2a591ca